components/infobox/extensions/commons/age.lua

```diff
@@ -0,0 +1,144 @@
+
+local _EPOCH_DATE = { year = 1970, month = 1, day = 1 }
+local _DEFAULT_DAYS_IN_MONTH = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 }
+local _MONTH_DECEMBER = 12
+local _CURRENT_YEAR = tonumber(mw.getContentLanguage():formatDate('Y'))
+
+---
+-- Represents a date
+--
+-- Before accessing the `year`, `month` and `day` values, please verify
+-- whether `isExact` or `isEmpty` are set to false or true respectively.
+--
+-- `isExact`: Whether this is a complete date, consisting of YYYY-MM-DD
+-- `isEmpty`: Whether there is a date here at all.
+-- `time`: Time in seconds
+local Date = Class.new(
+	function(self, dateString)
+		if String.isEmpty(dateString) then
+			self.isExact = false
+			self.isEmpty = true
+			return
+		end
+
+		self.isEmpty = false
+		self.isExact = true
+		local fields = mw.text.split(dateString, '-')
+
+		self.year = tonumber(fields[1])
+		self.month = tonumber(fields[2])
+		self.day = tonumber(fields[3])
+
+		if self.year == nil or self.month == nil or self.day == nil then
+			self.isExact = false
+		else
+			self.time = os.time({
+				year = self.year,
+				month = self.month,
+				day = self.day
+			})
+		end
+
+	end
+)
+
+function Date:getEarliestPossible()
+	return os.time({
+		year = self.year or _EPOCH_DATE.year,
+		month = self.month or _EPOCH_DATE.month,
+		day = self.day or _EPOCH_DATE.day,
+	})
+end
+
+function Date:getLatestPossible()
+	return os.time({
+		year = self.year or _CURRENT_YEAR,
+		month = self.month or _MONTH_DECEMBER,
+		day = self.day or _DEFAULT_DAYS_IN_MONTH[self.month or _MONTH_DECEMBER],
+	})
+end
+
+local BirthDate = Class.new(Date)
+local DeathDate = Class.new(Date)
+
+local Age = Class.new(
+	function(self, birthDate, deathDate)
+		self.birthDate = birthDate
+		self.deathDate = deathDate
+	end
+)
+
+function Age:calculate()
+	local endDate
+	if self.deathDate.isEmpty then
+		endDate = os.time()
+	else
+		endDate = self.deathDate.time
+	end
+
+	if self.birthDate.isExact and (self.deathDate.exact or self.deathDate.isEmpty) then
+		return self:_secondsToAge(os.difftime(endDate, self.birthDate.time))
+	elseif self.birthDate.year and (self.deathDate.year or self.deathDate.isEmpty) then
+		local minEndDate
+		local maxEndDate
+		if self.deathDate.isEmpty then
+			minEndDate = os.time()
+			maxEndDate = os.time()
+		else
+			minEndDate = self.deathDate:getEarliestPossible()
+			maxEndDate = self.deathDate:getLatestPossible()
+		end
+
+		local minAge = self:_secondsToAge(os.difftime(minEndDate, self.birthDate:getLatestPossible()))
+		local maxAge = self:_secondsToAge(os.difftime(maxEndDate, self.birthDate:getEarliestPossible()))
+
+		-- If our min and max age values are identical, then we can just display that value. Else,
+		-- we have to display a range of ages
+		if minAge == maxAge then
+			return minAge
+		else
+			return minAge .. '-' .. maxAge
+		end
+	end
+
+	return nil
+end
+
+function Age:_secondsToAge(seconds)
+	return math.floor(seconds / 60 / 60 / 24 / 365.25)
```

Comment:
does this account for all possible cases?
(like year dividable by 100, but not 400)
(value would be 365.2425 if considering that)
due to natural age limitations for people it probably is irrelevant though

Reply:
I am confident this works fine for all our cases